test/integration/gardener/reconcile/gardener_reconcile_test.go

```diff
@@ -0,0 +1,120 @@
+	"github.com/gardener/gardener/pkg/apis/garden/v1beta1"
+	"github.com/gardener/gardener/pkg/logger"
+	"github.com/gardener/gardener/pkg/utils/retry"
+	"github.com/gardener/gardener/test/integration/framework"
+	"github.com/sirupsen/logrus"
+	"time"
+
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	. "github.com/gardener/gardener/test/integration/shoots"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var (
+	kubeconfigPath  = flag.String("kubeconfig", "", "the path to the kubeconfig path  of the garden cluster that will be used for integration tests")
+	gardenerVersion = flag.String("version", "", "current gardener version")
+	logLevel        = flag.String("verbose", "", "verbosity level, when set, logging level will be DEBUG")
+
+	testLogger *logrus.Logger
+)
+
+const (
+	ReconcileShootsTimeout = 1 * time.Hour
+	InitializationTimeout  = 20 * time.Second
+)
+
+func validateFlags() {
+
+	if !StringSet(*kubeconfigPath) {
+		Fail("you need to specify the correct path for the kubeconfigpath")
+	}
+
+	if !FileExists(*kubeconfigPath) {
+		Fail("kubeconfigpath path does not exist")
+	}
+
+	if !StringSet(*gardenerVersion) {
+		Fail("you need to specify the current gardener version")
+	}
+}
+
+var _ = Describe("Shoot reconciliation testing", func() {
+	var (
+		gardenClient kubernetes.Interface
+	)
+
+	CBeforeSuite(func(ctx context.Context) {
+		validateFlags()
+		testLogger = logger.AddWriter(logger.NewLogger(*logLevel), GinkgoWriter)
+
+		var err error
+		gardenClient, err = kubernetes.NewClientFromFile("", *kubeconfigPath, client.Options{
+			Scheme: kubernetes.GardenScheme,
+		})
+		Expect(err).ToNot(HaveOccurred())
+
+	}, InitializationTimeout)
+
+	CIt("Should reconcile all shoots", func(ctx context.Context) {
+		err := retry.UntilTimeout(ctx, 30*time.Second, ReconcileShootsTimeout, func(ctx context.Context) (bool, error) {
+			shoots := &v1beta1.ShootList{}
+			err := gardenClient.Client().List(ctx, shoots)
+			if err != nil {
+				testLogger.Debug(err.Error())
+				return retry.MinorError(err)
+			}
+
+			// Directly return if the cluster does not contain any shoots
+			if len(shoots.Items) == 0 {
+				testLogger.Info("no shoots to reconcile")
+				return retry.Ok()
+			}
+
+			reconciledShoots := 0
+			for _, shoot := range shoots.Items {
+				// check if the last acted gardener version is the current version,
+				// to determine if the updated gardener version reconciled the shoot.
+				if shoot.Status.Gardener.Version != *gardenerVersion {
+					testLogger.Debugf("last acted gardener version %s does not match current gardener version %s", shoot.Status.Gardener.Version, *gardenerVersion)
+					continue
+				}
+				if framework.ShootCreationCompleted(&shoot) {
+					reconciledShoots++
+				}
+			}
+
+			if reconciledShoots != len(shoots.Items) {
+				testLogger.Infof("Reconciled %d of %d shoots. Waiting ...", reconciledShoots, len(shoots.Items))
+				return retry.NotOk()
```

Comment:
Try not to use `NotOk` - rather do a `MinorError` with the message you have in your `Infof`